Code Refactor: pursue extendability and readability

saart opened the pull request.

saart commented on netshare/models/doppelganger_tf/dataset.py:
```diff
@@ -1,19 +1,18 @@
+import copy
```
Note: nothing has been changed in this file besides formating and imports

saart commented on netshare/models/doppelganger_tf/util.py:
```diff
@@ -1,45 +1,24 @@
-from netshare.utils import OutputType, Output, Normalization
```
Note: nothing has been changed in this file besides formating and imports

saart commented on netshare/models/doppelganger_tf_model.py:
```diff
@@ -1,31 +1,35 @@
 import inspect
```
Note: nothing has been changed in this file besides formating and imports

fjxmlzn commented on netshare/utils/field.py:
```diff
@@ -32,42 +40,54 @@ def __init__(self, norm_option, min_x=None, max_x=None, dim_x=1,
         elif self.norm_option == Normalization.MINUSONE_ONE:
-            return np.asarray(2 * (x - self.min_x)
-                              / (self.max_x - self.min_x) - 1)
+            if self.max_x - self.min_x - 1 == 0:
```
Why there is "-1"?

saart replied:
You're right here, that was my miss. I tried to avoid `ZeroDivisionError`. Will be fixed in a sec!

saart commented on examples/sensors/config_example_sensors.json:
```diff
@@ -0,0 +1,94 @@
+                "encoding": "categorical"
+            }
+        ],
+        "timeseries": [
```
Use all fields from the original data

saart commented on examples/sensors/config_example_sensors.json:
```diff
@@ -0,0 +1,94 @@
+{
```
Move the sensors example out of the refactor branch

saart commented on README.md:
```diff
@@ -34,7 +34,7 @@ You may also refer to the [README](traces/README.md) for detailed descriptions o
 
```
put the diagram figure here

jinminhao commented on netshare/learn/model_learn.py:
```diff
@@ -0,0 +1,62 @@
+) -> None:
+    """
+    This function train the given config group and list of configs.
+    We first train the chunk0 model, then train the rest of the models TODO: why?
```
Suppose we have N+1 chunks of data, then we will have N+1 models.

With no DP, we would first train a model given the first chunk of data (D_chunk0). Then the rest of the models will be trained by using the first trained model as a base and fine-tuning the parameters(weights) given the different chunks of data(D_chunk1 - D_chunkN).

With DP, we would first train a pre-model using a public dataset. Then the N+1 models will be trained by fine-tuning the pre-model given data(D_chunk0 - D_chunkN) while DP-SGD is being activated.

The above explanation is a simplified version where we only setup one set of hyper-parameters. In reality, we would setup N sets (N > 1) since we have no idea which one is the best. Therefore, each chunk of data will have  N trained models. After the model generation, we would pick the best model given the metrics for each chunk of data. I tried to explain in details in order to avoid any potential confusion related to my another comments. Sorry for the heavy text:)

saart replied:
Ohh that's a wonderful explanation, that you!! :)
Just to make sure - 
1. currently, does the code assume one set of hyper-parameters? I.e. when you say "in reality..." do you mean "in future implementations.."?
2. in DP, where does the public dataset come from? Do we assume that this is chunk0? 
I updated the documentation, WDYT?

sniperyyc replied:
1. The current code assumes **multiple sets** of hyperparameters (to the best of my knowledge).
2. The public datasets come from public sources -- if you check the NetShare paper, e.g., when private dataset is CAIDA, public dataset can be CAIDA or DC. Public dataset is not necessarily chunk0 since all chunks will be trained on the public model. We did not include the DP version in the current branch (we use a two-step manual fashion instead for now). We agreed earlier that DP version is less prioritized since most ppl care about non-DP cases and DP cases would require some other minor tweaks to the code which we did not include in the current version. We could work on this feature later when we have a stable version IMO. **The updated doc looks great.** Thank you!

saart replied:
Ohhh now it makes sense! I couldn't find the logic of the DP, so I was confused!
Cool, thank you so much!
I think that I'll update the doc to explicitly say that we don't support DP now. It will solve future misunderstanding, ok?

sniperyyc replied:
Sounds great, Saar!

(We actually already included some DP-related code snippets (e.g., configurations like `pretrain_dir`, `dp_noise_multiplier` which you may already see in the refactoring) and we can do another dp-specific integration later as we will be using the torch-specific DP library anyway in the future instead of the outdated old-version tf-privacy :))

jinminhao commented on netshare/generate/generate.py:
```diff
@@ -0,0 +1,26 @@
+    copy_files(chosen_data_dir, get_generated_data_dir())
+
+
+def choose_best_chunk(denormalized_fields_dir: str) -> str:
```
I personally think the function name should not be `choose_best_chunk` because for each chunk of data, we will generate several models with different hyper-parameters. This function is to pick the best model for each chunk.

saart replied:
You're absolutely right, I'll rename to: `choose_best_model`

jinminhao commented on netshare/generate/denormalize_fields.py:
```diff
@@ -0,0 +1,111 @@
+            writer.writerow(session_data + timeseries_data)
+
+
+def denormalize_fields() -> str:
```
Timestamp may have some logical difference with other features in `timeseries`. For example, when the user is specifying "interarrival", we are calculating the delta of timestamp between two consecutive time steps in the same session as a fields. Therefore, when we are denormalizing the generated data, we should group by session key and also accumulate that field in order to get its absolute timestamp.

saart replied:
hmm that's a really good point!
It seems that I missed that in my implementation, right?
can you refer my to the code in the old implementation, so I can double-check that I have everything?
If I understand correctly, the piece of code that you refer to is in `merge_attr` inside `model_generate`, and it's not here in the `denormalize_fields`.
If this is indeed what I think it is, and you think that it makes more sense, then I'll move it to here 👍

jinminhao replied:
I think it is previously in the function `denormalize`. After we finish model generating, we need to merge all the data across chunks together and do `denormalization`. In the latest version of code in the branch new_dataset, we just commented out the `denormalization` since we haven't finished it for the generic dataset. We can also add it quickly in this code refactoring version.

jinminhao commented:
Thank you so much for your great effort!!! Besides my comments, LGTM~

fjxmlzn commented on netshare/configs/__init__.py:
```diff
@@ -0,0 +1,51 @@
+    _config = config
+
+
+def get_config(
```
I understand that `path2` is for providing an alternative path in the config for supporting the use case where a config may appear either in global_config or its own section. But this solution (adding a parameter `path2` and an ad-hoc `if path2:`) is not very elegant. Also, it's hard to use as now users need to specify two parameters "path" and "path2". In addition, if a parameter was originally in its own section, and somehow later on we need to share it with other sections and therefore move it to `global_config` section, we will need to modify all places where we get this config.

One alternative is that when we load the config, we populate the configs in `global` section into all other sections. We can then remove `path2`, and when we want to `get_config`, we always use the path in its own section. One downside of this approach is that we do not have any isolation for the configs in `global` section (i.e., global config may be populated into sections that do not need it). But given that we do not have isolation in the current version anyways (all functions can access any part of the configs, including the ones in `global`), I think it is fine for now.

fjxmlzn replied:
I see there is another use case of `path2`, which is for specifying multiple config path options (e.g., for backward compatibility). Even in that case, I think it is still not elegant to use `path2` to support that. What if we have a third possibility? Definitely, we do not want to add another `path3`. 

For this use case, we can just manually use multiple calls of `get_config`. If we really want to accomplish it in one call, we can make `path` either a string or a list of strings (each element in the list is one possibility of the path). I would prefer the first option to make the API cleaner.

saart replied:
That's an interesting discussion!
I added `path2` only to allow alternative paths to find the relevant configuration (as you thought in the second comment).
I choose to use `path2` and not `List[str]` because I added it only for backward compatibility (i.e. that old configurations will still work on the current branch).
I agree that this is not fully clear. Here are my two suggestions:
1. We can change the naming here to make it more clear. WDYT about changing the name of `path2` to `backward_compatible_path`?
2. Alternatively, I prefer to change it to a general list of paths (instead of asking the user to call this function multiple times). I think that it will be a cleaner API / easier to use.

I'm really against moving configuration during runtime - it makes the flow much harder to follow. I.e. when a component wants to use the configuration, it doesn't know where to search it, because something has moved it around. It is much much easier to assume that the input config is *exactly* what you defined in the config file.

fjxmlzn replied:
Got it. That makes sense to me now! Thanks for the explanations! Both approaches you suggested sound great to me! I would probably slightly prefer the second one, just in case in the future we update the config format again and the code will need to support more than 3 possibilities.

saart replied:
ok, then the second one it is :)

fjxmlzn commented on netshare/api.py:
```diff
@@ -0,0 +1,41 @@
+        ray.init(address="auto")
+
+    def train(self) -> None:
+        input_adapter()
```
Nitpicks: `generate` and `learn` are verbs, but `input_adapter`, `output_adapter` are nouns. They are just different steps in the pipeline, so maybe it is better to use the same form of words to name them to reduce confusion

saart replied:
You're absolutely right!
Love the idea!
Will do :)

fjxmlzn commented on netshare/learn/utils/dataframe_utils.py:
```diff
@@ -0,0 +1,90 @@
+    This function load the CSV files into a pandas dataframe.
+    """
+    dfs = []
+    for filename in os.listdir(csv_dir):
```
It might be better to do file extension filtering (it's in canonical format, so we know the extension). The reason is that MacOS/Windows might create some hidden files (e.g., . DS_Store). We don't want to read them and get into errors.

fjxmlzn commented on netshare/configs/__init__.py:
```diff
@@ -0,0 +1,51 @@
+from .default import __path__ as defaults_path
+
+_config: Optional[Config] = None
+
```
The default configs in `netshare/configs/default` are still in the old format. I understand that part of the code in this PR is backward compatible with the old format. But we might still want to update the default configs in this PR so that users will be inheriting the new config standards instead of the old ones. Also, the config readme `netshare/configs/README.md` should be updated. Especially, some outdated links are there.

saart replied:
You're abs right. Will do 👌

fjxmlzn commented on netshare/learn/utils/dataframe_utils.py:
```diff
@@ -0,0 +1,90 @@
+                & (big_raw_df[time_col_name] < time_evenly_spaced[chunk_id + 1])
+            ]
+            if len(df_chunk) == 0:
+                print("Raw chunk_id: {}, empty df_chunk!".format(chunk_id))
```
Use logger

fjxmlzn commented on netshare/utils/field.py:
```diff
@@ -148,14 +176,28 @@ def denormalize(self, bin_x):
+
+
+def field_config_to_key(field: Config) -> FieldKey:
+    return (
```
Saar suggested using namedtuple before. I think that's a good idea. Is there any particular reason why we do not use it here?

saart replied:
ohh because this object is not aimed to "hold" data, but only as a key to a hashmap.
I don't think that people should use the content of the object that is returned here 🤔

fjxmlzn replied:
Got it. That makes sense! Thanks for the explanations!

fjxmlzn commented on netshare/learn/setup_cross_chunks_data.py:
```diff
@@ -0,0 +1,256 @@
+            )
+        )
+
+    field_name = getattr(field, "name", "") or field.get("column", "") or field.columns
```
Is there a particular reason why we mix the usage of the `getattr` API and `.get` API of ConfigIO? If both work, we can just use one API to make it consistent?

saart replied:
Programmers should actually use them differently: `getattr` aimed to grab an attribute from an object, and `get` grabs a key from a dictionary!
This is `Config`'s dark magic that supports both of them 😂 
I'll change to `.get`, so it will be consistent as you suggested (usually the use of `getattr` is considered a smell)  :)

saart replied:
I'll also change things like `field.type` to `field["type"]`. I think that it also helps the reader understand that there's no complex object here, only fetching data from the config.

fjxmlzn commented on netshare/learn/setup_cross_chunks_data.py:
```diff
@@ -0,0 +1,256 @@
+
+    # Categorical field: (string | integer)
+    elif "categorical" in getattr(field, "encoding", "") or field.type == "string":
+        if field.type not in ["string", "integer"]:
```
Maybe it is a good idea to make all these constant strings (e.g., "string", "integer", "word2vec") in a separate constant.py (or in some other better structures), and all the referrals of those strings are using the constant variables instead of typing these strings manually and repeatedly. This would avoid the issues of typos in the strings (we would get NameError instead of silent bugs).

But maybe we can do this later, not in this PR.

saart replied:
hmm that's an interesting suggestion.
I agree that sometimes having a `constants` file is very good. But IMO it makes sense only when the strings should be shared between multiple files. This specific example is a configuration parsing that happens just once, therefore I don;t think that we should extract these strings.
But, I'll take this comment in the general context, and will do another pass on the whole PR and will look for constants that

fjxmlzn commented on netshare/learn/setup_per_chunk_data.py:
```diff
@@ -0,0 +1,384 @@
+    #   formats between the different PrePostProcessor
+    should_sample = (
+        get_config("pre_post_processor.class", default_value="")
+        == "DGRowPerSamplePrePostProcessor"
```
Is there a better way than hard-coding the class name here?

saart replied:
As I wrote in the TODO - this is pretty hard while maintaining backward compatibility.
Given that I'm rewriting the config files, I'll break the backward compatibility only in this specific case 🤷

fjxmlzn replied:
Under the new config format, how should specify if we want to do the sampling?

fjxmlzn commented on netshare/learn/setup_per_chunk_data.py:
```diff
@@ -0,0 +1,384 @@
+    learn_api.create_dirs(chunk_id)
+    learn_api.write_raw_chunk(df_per_chunk, chunk_id)
+    learn_api.write_data_train_npz(
+        data_attribute,
```
When calling functions with many parameters like this, maybe it is more readable (and avoids bugs) by passing parameters using key=value fashion?

fjxmlzn commented on netshare/utils/model_configuration.py:
```diff
@@ -1,120 +1,138 @@
+        }
+    )
 
     # TODO: add preprocessing logic for DoppelGANger (single-chunk?)
```
Is it already done? The version before refactoring supports DoppelGANger native single-chunk datasets (by a dedicated ModelManager). It would be better if the refactored code still supports them. Otherwise, the refactored code has reduced functionality than the original code.

saart replied:
This is a comment from the original code, most of this code in this file hasn't been changed :)
Generally, I only moved code around and added logic, so (assuming that I didn't mess somewhere) we shouldn't have reduced functionality 👍 
I validated it by running manually the `tests/driver.py` and `examples/dg_table_row_per_sample/driver.py` useacases and validated that the output is the same.
Do you think that there's another usecase that I should check?

saart replied:
btw, you can see in GitHub merged view of the PR what was changed/added (green)/removed (red), so it is easy to track changes:
<img width="683" alt="image" src="https://user-images.githubusercontent.com/13601452/213143588-1a868e99-2e27-45dd-abb0-0f01b18eaaea.png">

fjxmlzn replied:
Thank you! In the previous version, for `examples/dg_table_row_per_sample` we used DG's pipeline in https://github.com/netsharecmu/NetShare/blob/42c438ee7cd0b50c99653182f88601f8540e3b93/netshare/model_managers/dg_model_manager.py. I think the current code is modified from NetShare's pipeline https://github.com/netsharecmu/NetShare/blob/42c438ee7cd0b50c99653182f88601f8540e3b93/netshare/model_managers/netshare_manager/netshare_manager.py. NetShare's pipeline is a superset of DG's so the end goal was to keep only NetShare's pipeline to support both use cases (i.e., what we are currently doing in the PR), but I am just not sure if those TODO has already been done or not. Maybe @sniperyyc can help confirm that. If they are done, we can remove those to-dos.

sniperyyc replied:
I will take care of those TODOs and test the new code for 1. wiki 2. pcap 3. netflow use cases.

fjxmlzn commented on netshare/generate/model_generate.py:
```diff
@@ -1,89 +1,117 @@
+            for config in configs
+        ]
+    )
+    time.sleep(10)
```
Why do we need the sleep?

saart replied:
I don't know, it was in the original code!
I thought that it might have an internal logic that I'm missing 😂 
As long as we don't have good tests, I don't feel comfortable with deleting parts that I don't understand. Now that you highlighted it, I'll try to remove it and validate manually that everything is ok

fjxmlzn replied:
Got it! Then maybe we can just leave it there for now until @sniperyyc helps double-check it. Probably there is something I missed too 😂

jinminhao replied:
Previously, sleep is used to help ray to sync the jobs. I understand that ray has mechanism which can synchronize distributed jobs as needed. But from my personal experiences, there's possibility (not large) that ray would not work as expected. So i just add `sleep` even though it's not elegent. We can also check if we can run correctly without `sleep` when using the latest version of ray.

fjxmlzn commented on netshare/generate/generate.py:
```diff
@@ -0,0 +1,27 @@
+    copy_files(chosen_data_dir, get_generated_data_dir())
+
+
+def choose_best_model(denormalized_fields_dir: str) -> str:
```
Just double-check if I understand it correctly--Is this just a placeholder that is not yet implemented?

saart replied:
yes :)

fjxmlzn commented on netshare/generate/model_generate.py:
```diff
@@ -1,89 +1,117 @@
+    )
+    time.sleep(10)
+
+    # TODO: The below part didn't happen in dg. Why?
```
DG considers a single chunk so we do not need to do this. This technique was introduced in NetShare paper. In general, if the user specifies to use a single chunk, we may just use the DG processing flow and many steps (including this) can be skipped.

saart replied:
ohh cool, so I'll put it inside a `if n_chunk > 1` clause 💪

jinminhao commented:
Hey Saar, since the code-refactoring version of netshare hasn't included a detailed implementation of `denormalize`, do you mind I working on this branch to add the related logic? Or do you have any better idea on that?

saart commented:
> Hey Saar, since the code-refactoring version of netshare hasn't included a detailed implementation of `denormalize`, do you mind I working on this branch to add the related logic? Or do you have any better idea on that?

Hi @jinminhao I think that it is best that you will create another branch *from* this branch, and don't work directly on the same branch.
It will be much much easier to maintain everything in that order: it will be clear what is the change that you made, and we can merge both things independently :)

jinminhao commented:
> 

Sounds great!

jinminhao commented:
> > Hey Saar, since the code-refactoring version of netshare hasn't included a detailed implementation of `denormalize`, do you mind I working on this branch to add the related logic? Or do you have any better idea on that?
> 
> Hi @jinminhao I think that it is best that you will create another branch _from_ this branch, and don't work directly on the same branch. It will be much much easier to maintain everything in that order: it will be clear what is the change that you made, and we can merge both things independently :)

Sounds great. I will try to finish it before this weekend. thx!

sniperyyc commented:
> > Hey Saar, since the code-refactoring version of netshare hasn't included a detailed implementation of `denormalize`, do you mind I working on this branch to add the related logic? Or do you have any better idea on that?
> 
> Hi @jinminhao I think that it is best that you will create another branch _from_ this branch, and don't work directly on the same branch. It will be much much easier to maintain everything in that order: it will be clear what is the change that you made, and we can merge both things independently :)

I will also create a new branch from this branch to add the pcap example and do in-line fixes as I see needed. I will also test the new code for 1. wiki 2. pcap 3. netflow datasets for sanity checks (i.e., making sure we are generating good-quality traces).

sniperyyc commented on netshare/learn/setup_per_chunk_data.py:
```diff
@@ -0,0 +1,385 @@
+        and generate the normalized timestamp column if needed.
+
+    There are two types of timestamp generation:
+    1. encoding = raw: Taking a specific column as another timestamp column. TODO: I don't understand why this is needed.
```
It's just one method we have tried in the past :)

saart replied:
I see,
Then I'll delete it if this is ok with you.
I prefer to no leave unused code in the codebase :)

sniperyyc replied:
I may choose to keep it as it gives another option to the user to encode the timestamp.

sniperyyc commented on netshare/learn/setup_per_chunk_data.py:
```diff
@@ -0,0 +1,385 @@
+        and also take the diff between every packet to the previous one as a feature column.
+    """
+    session_key_config = get_config(
+        ["pre_post_processor.config.metadata", "learn.session_key"]
```
Are `learn.session_key`, `learn.timestamp` related new configs uploaded somewhere? I understand that the `pre_post_processor.config.metadata` is mainly for backward compatability.

saart replied:
Yes. I thought that I updated all the configs.
As an example, you can look on `dg_table_row_per_sample/config_example_wiki.json`

sniperyyc replied:
I see. Is netflow/pcap related configs rewritten to match the `learn.session_key` and `learn.timeseries`?

sniperyyc replied:
I could help update the remaining netflow/pcap configs if that's not done yet -- another good exercise for me to get more familiar with the new code :)

saart replied:
I just checked: I did it completely in `netshare/configs`, but have missed some in `examples/`.
If you want to do it, then it will be awsome. If not, I can do it as well! :D

sniperyyc replied:
Cool, thanks for letting me know! 
I think the current [pcap default config](https://github.com/netsharecmu/NetShare/blob/code-refactor/netshare/configs/default/pcap.json) and [netflow default config](https://github.com/netsharecmu/NetShare/blob/code-refactor/netshare/configs/default/netflow.json) are partially updated (as I did not see `learn.session_key` or `learn.timeseries` in these two files and probably other latest config keys)
Sure, I will update these configs in my PR. Thank you!!

sniperyyc commented on netshare/generate/denormalize_fields.py:
```diff
@@ -0,0 +1,138 @@
+    has ended in this time step.
+    """
+    os.makedirs(csv_folder, exist_ok=True)
+    csv_path = os.path.join(csv_folder, f"data_{filename}_{random.random()}.csv")
```
Is there any specific concern to include `random.random()` in the filename?

saart replied:
yes. I was afraid that `filename` is not unique enough, thus we will override ourselves.
Checking it retrospectively, I think that this is not necessary, and we can remove it if your prefer

sniperyyc commented:
A few qns on the result folder/files structure/contents. Here is the current result folder structure:
```
../results/test_ugr16-451882/
├── canonical_input
├── generated
├── logs
├── models
├── output_data
├── pre_processed_data
└── raw_generated
```

Q1: Is `logs` still in use? I saw partial logs remaining from previous architecture in `logs` and I believe we must have a better logging system and logging information now?

Q2: For each chunk's generated data, e.g., `output_data/sample_len-10/data_chunk_id-0_iteration_id-3_0.7727486694654094.csv`, the columns here are different from what we see in the raw data.
- The columns in the synthetic data are: `srcip,dstip,srcport,dstport,proto,startFromThisChunk,chunk_0,chunk_1,flow_start,td,pkt,byt,type,interarrival_within_flow`
- The columns in the real data are:
`srcip,dstip,srcport,dstport,proto,ts,td,pkt,byt,type`

*We should `denormalize` synthetic data to have the exact same format as real data. I could take care of this part.*

Q3: some other minor pts.

@jinminhao has helped fix most issues in the `postprocess` part. We sincerely apologize for not including the complete implementation in the `new_dataset` branch which causes a lot of confusion to you. The main branch includes [the logic that we would like to apply to the `postprocess` part](https://github.com/netsharecmu/NetShare/blob/800cb166024d0c5c0d0b4c8e3e845864e75cd93a/netshare/pre_post_processors/netshare/netshare_pre_post_processor.py#L453-L535). **I will quickly fix this part in the PR that I opened (probably excluding the picking best model part) and asked for CR by all of you.** Thanks!

sniperyyc commented:
> A few qns on the result folder/files structure/contents. Here is the current result folder structure:
> 
> ```
> ../results/test_ugr16-451882/
> ├── canonical_input
> ├── generated
> ├── logs
> ├── models
> ├── output_data
> ├── pre_processed_data
> └── raw_generated
> ```
> 
> Q1: Is `logs` still in use? I saw partial logs remaining from previous architecture in `logs` and I believe we must have a better logging system and logging information now?
> 
> Q2: For each chunk's generated data, e.g., `output_data/sample_len-10/data_chunk_id-0_iteration_id-3_0.7727486694654094.csv`, the columns here are different from what we see in the raw data.
> 
> * The columns in the synthetic data are: `srcip,dstip,srcport,dstport,proto,startFromThisChunk,chunk_0,chunk_1,flow_start,td,pkt,byt,type,interarrival_within_flow`
> * The columns in the real data are:
>   `srcip,dstip,srcport,dstport,proto,ts,td,pkt,byt,type`
> 
> _We should `denormalize` synthetic data to have the exact same format as real data. I could take care of this part._
> 
> Q3: some other minor pts.
> 
> @jinminhao has helped fix most issues in the `postprocess` part. We sincerely apologize for not including the complete implementation in the `new_dataset` branch which causes a lot of confusion to you. The main branch includes [the logic that we would like to apply to the `postprocess` part](https://github.com/netsharecmu/NetShare/blob/800cb166024d0c5c0d0b4c8e3e845864e75cd93a/netshare/pre_post_processors/netshare/netshare_pre_post_processor.py#L453-L535). **I will quickly fix this part in the PR that I opened (probably excluding the picking best model part) and asked for CR by all of you.** Thanks!

Except the `denormalize` part, other modules LGTM! I will lyk if I get into any issue during the manual test. Thanks for your heroic efforts!!

sniperyyc commented:
https://github.com/netsharecmu/NetShare/blob/cb41f33b5f5cbc52ec97ed928194f3e354d6fca5/netshare/configs/default/netflow.json#L18-L35

Is the `model_manager` kept in the config only for backward compatibility?

sniperyyc commented:
Empty output when `n_chunks=1`. Two examples to reproduce this error:

1. `examples/dg_table_row_per_sample/driver.py` (where `n_chunks=1`)
2. `tests/driver.py` and modify `n_chunks=1` in `examples/netflow/config_example_netflow_nodp.json`.

Will fix this along with the `denormalize` part.

sniperyyc commented on netshare/learn/setup_per_chunk_data.py:
```diff
@@ -0,0 +1,385 @@
+            drop=True
+        )  # reset index to make it start from zero
+        data_feature_list.append(df_group[new_timeseries_list].to_numpy())
+        attr_per_row = apply_cross_chunk_mechanism(
+            df_group=df_group, cross_chunks_data=cross_chunks_data, chunk_id=chunk_id
+        )
```
This may be incorrect -- `cross_chunk_data` contains the unnormalized metadata but we are passing the normalized metadata here. Will raise the `{metadata} not found` error? Will fix as needed.

saart replied:
I'm not sure that I follow - `cross_chunk_data` shouldn't hold any concrete data, but only "metadata" on how to execute the data manipulation in the per-chunk phase. 
E.g. it holds the word2vec model and the columns configuration, but it doesn't include any information from the data itself.

sniperyyc commented on netshare/learn/setup_cross_chunks_data.py:
```diff
@@ -0,0 +1,271 @@
+    flowkeys_chunkidx = get_flowkeys_chunkidx(df_chunks)
+    global_max_flow_len = get_global_max_flow_len(df_chunks)
+
+    return CrossChunksData(
+        embed_model=embed_model,
+        flowkeys_chunkidx=flowkeys_chunkidx,
+        global_max_flow_len=global_max_flow_len,
+        session_key_fields=session_key_fields,
+        timeseries_fields=timeseries_fields,
+    )
```
Are we storing `flowkeys_chunkdix` and other utility plots? These could be handy in terms of debugging/checking stats.

saart replied:
Hmm that's a great idea.
I think that it makes sense to log it in `debug`, thus it will be shown only when we try to debug the app, but it will not be printed in standard executions

saart commented:
> A few qns on the result folder/files structure/contents. Here is the current result folder structure:
> 
> ```
> ../results/test_ugr16-451882/
> ├── canonical_input
> ├── generated
> ├── logs
> ├── models
> ├── output_data
> ├── pre_processed_data
> └── raw_generated
> ```
> 
> Q1: Is `logs` still in use? I saw partial logs remaining from previous architecture in `logs` and I believe we must have a better logging system and logging information now?
> 
> Q2: For each chunk's generated data, e.g., `output_data/sample_len-10/data_chunk_id-0_iteration_id-3_0.7727486694654094.csv`, the columns here are different from what we see in the raw data.
> 
> * The columns in the synthetic data are: `srcip,dstip,srcport,dstport,proto,startFromThisChunk,chunk_0,chunk_1,flow_start,td,pkt,byt,type,interarrival_within_flow`
> * The columns in the real data are:
>   `srcip,dstip,srcport,dstport,proto,ts,td,pkt,byt,type`
> 
> _We should `denormalize` synthetic data to have the exact same format as real data. I could take care of this part._
> 
> Q3: some other minor pts.
> 
> @jinminhao has helped fix most issues in the `postprocess` part. We sincerely apologize for not including the complete implementation in the `new_dataset` branch which causes a lot of confusion to you. The main branch includes [the logic that we would like to apply to the `postprocess` part](https://github.com/netsharecmu/NetShare/blob/800cb166024d0c5c0d0b4c8e3e845864e75cd93a/netshare/pre_post_processors/netshare/netshare_pre_post_processor.py#L453-L535). **I will quickly fix this part in the PR that I opened (probably excluding the picking best model part) and asked for CR by all of you.** Thanks!

@sniperyyc 
Q1: I tried to remove all the prints messages, thus it shouldn't be in use. 
Q2: That's indeed a bug. So there are more columns than expected, right? I.e. we do insert the right columns in the output, but we also add irrelevant columns? Amazing catch!
Q3:  No worries! I'm looking forward to get your PR 🤩

saart commented:
> https://github.com/netsharecmu/NetShare/blob/cb41f33b5f5cbc52ec97ed928194f3e354d6fca5/netshare/configs/default/netflow.json#L18-L35
> 
> Is the `model_manager` kept in the config only for backward compatibility?

Nope. This is part of the configuration that I haven't changed.
I think that it's currently not worth it because there are many details that are relevant to TF1. IMO we should discuss it thoroughly when migrating to PyTorch :)

saart commented:
> Empty output when `n_chunks=1`. Two examples to reproduce this error:
> 
> 1. `examples/dg_table_row_per_sample/driver.py` (where `n_chunks=1`)
> 2. `tests/driver.py` and modify `n_chunks=1` in `examples/netflow/config_example_netflow_nodp.json`.
> 
> Will fix this along with the `denormalize` part.

I think that I solved it already. Can you please try to pull from origin? (E.g. execute `git pull origin code-refactor` into your branch

sniperyyc commented:
> > Empty output when `n_chunks=1`. Two examples to reproduce this error:
> > 
> > 1. `examples/dg_table_row_per_sample/driver.py` (where `n_chunks=1`)
> > 2. `tests/driver.py` and modify `n_chunks=1` in `examples/netflow/config_example_netflow_nodp.json`.
> > 
> > Will fix this along with the `denormalize` part.
> 
> I think that I solved it already. Can you please try to pull from origin? (E.g. execute `git pull origin code-refactor` into your branch

Sorry my bad, it works but inherits the same bug that multi-chunks mode has i.e., having additional columns not present in the original data. Will fix as well :)

sniperyyc commented:
Fixed in [this commit](https://github.com/netsharecmu/NetShare/pull/15/commits/87ce90ea2ef1cef53002e36a4646a7f9c88f485a): add error prompts for zero case, i.e., terminate the generation process

Reason: ok, figured out. When `num_iteration` is very small, i.e., the model has not learned anything, there is a chance that `merged attributes` will have zero valid values -- which makes the `generate_given_attr` takes empty array as input, which throws an error.

Update: will occasionally occur; potential cause in the generation part. I will take a stab.

Another caught bug: 

https://github.com/netsharecmu/NetShare/blob/cb41f33b5f5cbc52ec97ed928194f3e354d6fca5/examples/netflow/config_example_netflow_nodp.json#L95-L97

When `sample_len` has multiple options, e.g., change the above lines to

```
"sample_len": [1, 5, 10]
```
raises an error.

sniperyyc commented on netshare/generate/denormalize_fields.py:
```diff
@@ -0,0 +1,138 @@
+            session_data_per_session,
+            timeseries_per_session,
+        ) in zip(data_gen_flag, session_key, timeseries):
+            # this if is here in parallel to the if in `reduce_samples`. It supports old flows.
+            session_data_per_session = session_data_per_session.tolist()
+            for j in range(data_gen_per_session.shape[0]):
+                if data_gen_per_session[j] == 1.0:
+                    timeseries_data = timeseries_per_session[j].tolist()
+                    writer.writerow(session_data_per_session + timeseries_data)
```
Not sure what the comment (line 94) means here -- I thought line 97-99 are compulsory by default since not all flows are the same length?

saart replied:
This code is here just because it appears in the original logic.
I didn't understand it as well, and that's what the comment tries to say.
If I understand correctly, this code is relevant only for the "session-based dataset" (wiki) that we wanted to get rid of, but haven't got to it yet.
Does it create a bug in your flow?

sniperyyc replied:
Oh, nothing related to the bug and I think the code itself makes sense -- anyway we need `data_gen_flag` to extract the "valid" timeseries. I was just confused about the comments. Thanks for the explanation!

sniperyyc commented:
Is there a particular reason that we are using `mkdtemp` across the codebase w/o being cleaned up? It quickly eats up my disk space so I replaced all temp files creation in the `generate` module (mainly around `denormalize`).

IMO, we should use different native subfolders (e.g., `canonical_input`, `preprocessed_data`) to serve as input/output between modules instead of temp files :)

sniperyyc commented on netshare/utils/field.py:
```diff
@@ -1,73 +1,124 @@
+
         # [0, 1] normalization
         if self.norm_option == Normalization.ZERO_ONE:
+            if self.max_x - self.min_x == 0:
+                raise Exception("Not enough data to proceed!")
```
I think this error is due to "constant" values instead of "not enough data".

Constant values happen a lot and why not just add a small epsilon (e.g., 1e-8) to `self.max_x - self.min_x`?

saart replied:
Sounds absolutely correct!

saart commented:
> Is there a particular reason that we are using `mkdtemp` across the codebase w/o being cleaned up? It quickly eats up my disk space so I replaced all temp files creation in the `generate` module (mainly around `denormalize`).
> 
> IMO, we should use different native subfolders (e.g., `canonical_input`, `preprocessed_data`) to serve as input/output between modules instead of temp files :)

I used the `tempfile` API only for internal and temporary directories. It should be created in `/tmp`, which should be deleted frequently. If you prefer to use other locations and handle the deletion - sure, let's do it 👍 💪

sniperyyc commented:
Why are we replacing all native TF1 APIs with `tf.compat.v1`? This seems to be a big change and we should better sync up ahead of time regarding such changes ...

sniperyyc commented:
> > Is there a particular reason that we are using `mkdtemp` across the codebase w/o being cleaned up? It quickly eats up my disk space so I replaced all temp files creation in the `generate` module (mainly around `denormalize`).
> > IMO, we should use different native subfolders (e.g., `canonical_input`, `preprocessed_data`) to serve as input/output between modules instead of temp files :)
> 
> I used the `tempfile` API only for internal and temporary directories. It should be created in `/tmp`, which should be deleted frequently. If you prefer to use other locations and handle the deletion - sure, let's do it 👍 💪

I am fine with temp files as long as they get cleared after exit. Currently, on macOS the temp files will be at `/var/folders` and the temp files are NOT getting cleared automatically -- which leads to a huge usage of disk space.
Separately, I don't think we need temp files -- subfolders in the `worker_folder` could be properly used as inputs/outputs between different stages.

saart commented:
> Why are we replacing all native TF1 APIs with `tf.compat.v1`? This seems to be a big change and we should better sync up ahead of time regarding such changes ...

My goal in this change was to get reed from all the deprecation warnings on the *locations* of the functions, and moved to the new location.
I didn't think that it will change functionality, as I changed only locations. If there's any functionality change - please LMK and I'll rollback these changes ASAP!

saart commented:
> > > Is there a particular reason that we are using `mkdtemp` across the codebase w/o being cleaned up? It quickly eats up my disk space so I replaced all temp files creation in the `generate` module (mainly around `denormalize`).
> > > IMO, we should use different native subfolders (e.g., `canonical_input`, `preprocessed_data`) to serve as input/output between modules instead of temp files :)
> > 
> > 
> > I used the `tempfile` API only for internal and temporary directories. It should be created in `/tmp`, which should be deleted frequently. If you prefer to use other locations and handle the deletion - sure, let's do it 👍 💪
> 
> I am fine with temp files as long as they get cleared after exit. Currently, on macOS the temp files will be at `/var/folders` and the temp files are NOT getting cleared automatically -- which leads to a huge usage of disk space. Separately, I don't think we need temp files -- subfolders in the `worker_folder` could be properly used as inputs/outputs between different stages.

I see. You're right, I rechecked the [docs](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp), and indeed `"The user of mkdtemp is responsible for deleting the temporary directory and its contents when done with it.`. This is my fault!
So we can either delete it in the end, or use subfolders in the `worker_folder` as you suggested.
I'm absolutely fine with your suggestion! :)
Do you prefer that I'll do it now?

sniperyyc commented:
> > > > Is there a particular reason that we are using `mkdtemp` across the codebase w/o being cleaned up? It quickly eats up my disk space so I replaced all temp files creation in the `generate` module (mainly around `denormalize`).
> > > > IMO, we should use different native subfolders (e.g., `canonical_input`, `preprocessed_data`) to serve as input/output between modules instead of temp files :)
> > > 
> > > 
> > > I used the `tempfile` API only for internal and temporary directories. It should be created in `/tmp`, which should be deleted frequently. If you prefer to use other locations and handle the deletion - sure, let's do it 👍 💪
> > 
> > 
> > I am fine with temp files as long as they get cleared after exit. Currently, on macOS the temp files will be at `/var/folders` and the temp files are NOT getting cleared automatically -- which leads to a huge usage of disk space. Separately, I don't think we need temp files -- subfolders in the `worker_folder` could be properly used as inputs/outputs between different stages.
> 
> I see. You're right, I rechecked the [docs](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp), and indeed `"The user of mkdtemp is responsible for deleting the temporary directory and its contents when done with it.`. This is my fault! So we can either delete it in the end, or use subfolders in the `worker_folder` as you suggested. I'm absolutely fine with your suggestion! :) Do you prefer that I'll do it now?

No worries at all :) I would personally prefer the later one i.e., the use of subfolders in the `work_folder`. But if the previous one is easier for you and requires fewer changes, we can go with that.

Sure, please go ahead and edit as you like. If possible, would you mind working on top of `code-refactor-tests-diff` branch to minimize the conflict? Thank you! (You can take the lock -- I will not commit on that branch until you are done with the changes).

sniperyyc commented:
Merged all the proposed changes. 

Please feel free to continue making commits to `code-refactor` branch and open PR when CR is needed. Thanks!
